### ISSUE TITLE
[wiki] Tutorial

### ISSUE THREAD

**Issue by Jingkang50** [Jingkang50/OpenOOD#80]

**Comment by JediWarriorZou** [comment]
# 1.How to implement a method with train and test stage (With VOS as an example) ?
**`Train stage`:** 
- `Config`: You should write a few of configuration files with format ".yml". A configuration file is in "key-value" form. Generally, these configuration files are associated with five aspects.
  -  datasets: define details of your dataset including name, batch size, image size, number of class etc. In some case, multiple datasets  are used and splitted so that the configuration is rather complex.
  -  networks: define details of your network including  pretrain requirement, checkpoint of network etc. 
  -  pipelines: Certainly you will use a train pipeline. In this file, you can set the configuraion of your device such as number of worker and number of GPU. Moreover, you can set parameters of the trainer, evaluator, optimizer, scheduler, recorder etc.
  -  preprocessors: Some method need a special preprocessor. The base preprocessor is a default one.
  -  proprocessors: Some method need a special postprocessor. The base postprocessor is a default one.
  
- `Main code`:  It is challenging for you to implement the train stage in one python file. In the train stage, the code can be divided into 10 main sections. Some sections are optional which means your code may not require to amend these sections.
  - utils: It provides some functions including reading configuration files, generating logging files etc. You rarely need to amend this section.
  - losses: It defines a few of special and complicated loss functions.
  - datasets: It defines some dataset forms and provides two dataloader generators include "get_dataloader" and "get_ood_dataloader"  in utils.py. The most common dataset form is imglist. It generates datasets with respective txt files containing pathes of images and labels.You rarely need to amend this section.

**Comment by Zzitang** [comment]
# 2. Implement a method with test stage: take **Gram** as an example
1. For each approach, we suggest starting from the pipelines. In gram, we detect out-of-distribution samples by comparing each value in the gram matrix with its respective range observed over the training data. No training is required and we can directly employ the **TestOODPipeline**.
2. In TestOODPipeline, necessary components consist of dataloader, network, evaluator, and postprocessor. OPENOOD codebase provides an OODEvaluator suitable for most OOD Detection methods. Therefore, all we need is a postprocessor that provides methods to calculate necessary results like predictions and confidence.
3. Each postprocessor extends the BasePostprocessor class, which provides three basic method including setup(),  inference() and postprocess().
```python
class GRAMPostprocessor(BasePostprocessor):
    def __init__(self, config):
        self.config = config
        self.postprocessor_args = config.postprocessor.postprocessor_args
        self.num_classes = self.config.dataset.num_classes
        self.powers = self.postprocessor_args.powers

        self.feature_min, self.feature_max = None, None
```
- `setup()`:  The setup() method takes the input of both id and ood dataloaders to provide the necessary preparation for the upcoming process. 
	Here, as we know from the Gram method, we require the range of gram matrices of the training data, in particular, the maximum and minimum for gram matrices of each class, each layer of the network, and each order of gram matrices. In **GRAMPostprocessor**, we use `sample_estimator()` to calculate the required ranges of the feature using the model and train data.
```python
    def setup(self, net: nn.Module, id_loader_dict, ood_loader_dict):

        self.feature_min, self.feature_max = sample_estimator(
            net, id_loader_dict['train'], self.num_classes, self.powers)
```
```python
@torch.no_grad()
def sample_estimator(model, train_loader, num_classes, powers):

    model.eval()

    num_layer = 5
    num_poles_list = powers
    num_poles = len(num_poles_list)
    feature_class = [[[None for x in range(num_poles)]
                      for y in range(num_layer)] for z in range(num_classes)]
    label_list = []
    mins = [[[None for x in range(num_poles)] for y in range(num_layer)]
            for z in range(num_classes)]
    maxs = [[[None for x in range(num_poles)] for y in range(num_layer)]
            for z in range(num_classes)]

    # collect features and compute gram metrix
    for batch in tqdm(train_loader, desc='Compute min/max'):
        data = batch['data'].cuda()
        label = batch['label']
        _, feature_list = model(data, return_feature_list=True)
        label_list = tensor2list(label)
        for layer_idx in range(num_layer):

            for pole_idx, p in enumerate(num_poles_list):
                temp = feature_list[layer_idx].detach()

                temp = temp**p
                temp = temp.reshape(temp.shape[0], temp.shape[1], -1)
                temp = ((torch.matmul(temp,
                                      temp.transpose(dim0=2,
                                                     dim1=1)))).sum(dim=2)
                temp = (temp.sign() * torch.abs(temp)**(1 / p)).reshape(
                    temp.shape[0], -1)

                temp = tensor2list(temp)
                for feature, label in zip(temp, label_list):
                    if isinstance(feature_class[label][layer_idx][pole_idx],
                                  type(None)):
                        feature_class[label][layer_idx][pole_idx] = feature
                    else:
                        feature_class[label][layer_idx][pole_idx].extend(
                            feature)
    # compute mins/maxs
    for label in range(num_classes):
        for layer_idx in range(num_layer):
            for poles_idx in range(num_poles):
                feature = torch.tensor(
                    np.array(feature_class[label][layer_idx][poles_idx]))
                current_min = feature.min(dim=0, keepdim=True)[0]
                current_max = feature.max(dim=0, keepdim=True)[0]

                if mins[label][layer_idx][poles_idx] is None:
                    mins[label][layer_idx][poles_idx] = current_min
                    maxs[label][layer_idx][poles_idx] = current_max
                else:
                    mins[label][layer_idx][poles_idx] = torch.min(
                        current_min, mins[label][layer_idx][poles_idx])
                    maxs[label][layer_idx][poles_idx] = torch.max(
                        current_min, maxs[label][layer_idx][poles_idx])

    return mins, maxs
```
- `inference()`: The inference() method provides the predictions, confidence, and original labels for the input dataloader, which will be utilized in evaluators. This method employs the postproces() method to calculate the prediction and confidence for each data. Usually, there’s no need for custom postprocessors to rewrite the inference() method. Therefore, in GRAMPostprocessor, we only rewrite the setup() and postprocess() methods.
- `postprocess()`: The postprocess() method calculates the prediction and confidence for each input data.   
	For gram, we use get_deviations() which calculates the deviation of the gram metric of each data using the mins and maxs prepared in setup(). The deviations are directly employed as the confidence.
```python
    def postprocess(self, net: nn.Module, data: Any):
        preds, deviations = get_deviations(net, data, self.feature_min,
                                           self.feature_max, self.num_classes,
                                           self.powers)
        return preds, deviations
```
```python
def get_deviations(model, data, mins, maxs, num_classes, powers):
    model.eval()

    num_layer = 5
    num_poles_list = powers
    exist = 1
    pred_list = []
    dev = [0 for x in range(200)]

    # get predictions
    logits, feature_list = model(data, return_feature_list=True)
    confs = F.softmax(logits, dim=1).cpu().detach().numpy()
    preds = np.argmax(confs, axis=1)
    predsList = preds.tolist()
    preds = torch.tensor(preds)

    for pred in predsList:
        exist = 1
        if len(pred_list) == 0:
            pred_list.extend([pred])
        else:
            for pred_now in pred_list:
                if pred_now == pred:
                    exist = 0
            if exist == 1:
                pred_list.extend([pred])

    # compute sample level deviation
    for layer_idx in range(num_layer):
        for pole_idx, p in enumerate(num_poles_list):
            # get gram metirx
            temp = feature_list[layer_idx].detach()
            temp = temp**p
            temp = temp.reshape(temp.shape[0], temp.shape[1], -1)
            temp = ((torch.matmul(temp, temp.transpose(dim0=2,
                                                       dim1=1)))).sum(dim=2)
            temp = (temp.sign() * torch.abs(temp)**(1 / p)).reshape(
                temp.shape[0], -1)
            temp = tensor2list(temp)

            # compute the deviations with train data
            for idx in range(len(temp)):
                dev[idx] += (F.relu(mins[preds[idx]][layer_idx][pole_idx] -
                                    sum(temp[idx])) /
                             torch.abs(mins[preds[idx]][layer_idx][pole_idx] +
                                       10**-6)).sum()
                dev[idx] += (F.relu(
                    sum(temp[idx]) - maxs[preds[idx]][layer_idx][pole_idx]) /
                             torch.abs(maxs[preds[idx]][layer_idx][pole_idx] +
                                       10**-6)).sum()
    conf = [i / 50 for i in dev]

    return preds, torch.tensor(conf)
```
